File: codeforces/401/2c.rs
#[allow(unused_imports)]
use std::cmp::*;
#[allow(unused_imports)]
use std::collections::*;
use std::io::Read;
#[allow(dead_code)]
fn getline() -> String {
    let mut ret = String::new();
    std::io::stdin().read_line(&mut ret).ok().unwrap();
    ret
}
fn get_word() -> String {
    let mut stdin = std::io::stdin();
    let mut u8b: [u8; 1] = [0];
    loop {
        let mut buf: Vec<u8> = Vec::with_capacity(16);
        loop {
            let res = stdin.read(&mut u8b);
            if res.unwrap_or(0) == 0 || u8b[0] <= b' ' {
                break;
            } else {
                buf.push(u8b[0]);
            }
        }
        if buf.len() >= 1 {
            let ret = String::from_utf8(buf).unwrap();
            return ret;
        }
    }
}

#[allow(dead_code)]
fn get<T: std::str::FromStr>() -> T { get_word().parse().ok().unwrap() }

/**
 * Sparse Table.
 * BiOp should be the type of a binary operator which is
 * associative, commutative and idempotent.
 * (For example, both min and gcd satisfy these properties.)
 * Verified by: AtCoder CODE FESTIVAL 2016 Tournament Round 3 (Parallel) B
 * (http://cf16-tournament-round3-open.contest.atcoder.jp/submissions/1026294)
 */
struct SparseTable<T, BiOp> {
    biop: BiOp,
    st: Vec<Vec<T>>,
}

impl<T, BiOp> SparseTable<T, BiOp>
    where BiOp: Fn(T, T) -> T,
          T: Copy {
    pub fn new(ary: &[T], biop: BiOp) -> Self {
        let n = ary.len();
        let mut h = 1;
        while 1 << h < n {
            h += 1;
        }
        let mut st: Vec<Vec<T>> = vec![Vec::from(ary); h + 1];
        for i in 0 .. n {
            st[0][i] = ary[i];
        }
        for b in 1 .. (h + 1) {
            if n + 1 < 1 << b {
                break;
            }
            for i in 0 .. (n + 1 - (1 << b)) {
                let next_idx = (1 << (b - 1)) + i;
                st[b][i] = biop(st[b - 1][i], st[b - 1][next_idx]);
            }
        }
        SparseTable {biop: biop, st: st}
    }
    fn top_bit(t: usize) -> usize {
        let mut h = 0;
        while 1 << h <= t {
            h += 1;
        }
        h - 1
    }
    pub fn query(&self, f: usize, s: usize) -> T {
        assert!(f <= s);
        let b = Self::top_bit(s + 1 - f);
        let endpoint = s + 1 - (1 << b);
        (self.biop)(self.st[b][f], self.st[b][endpoint])
    }
}


fn solve() {
    let n = get();
    let m = get();
    let a: Vec<Vec<i64>> = (0 .. n).map(|_|
                         (0 .. m).map(|_| get::<i64>()).collect())
        .collect();
    let k = get();
    // We use either O(n^2 * m) or O(n * log(n) + k * m) algorithm,
    // depending on m.
    if m >= 320 {
        // Use O(n^2 * m) one.
        let mut tbl: Vec<Vec<Vec<bool>>> = vec![Vec::new(); n];
        for i in 0 .. n {
            let mut ok = vec![true; m];
            for j in i + 1 .. n {
                for c in 0 .. m {
                    ok[c] &= a[j][c] >= a[j - 1][c];
                }
                tbl[i].push(ok.clone());
            }
        }
        for _ in 0 .. k {
            let l: usize = get();
            let r: usize = get();
            let mut ok = false;
            if l < r {
                for &b in tbl[l - 1][r - l - 1].iter() {
                    ok |= b;
                }
            } else {
                ok = true;
            }
            println!("{}", if ok { "Yes" } else { "No" });
        }
    } else {
        // Use O(n * log(n) + k * m) one.
        let mut spts = Vec::new();
        for i in 0 .. m {
            let mut c = Vec::new();
            for j in 0 .. n - 1 {
                c.push(a[j + 1][i] >= a[j][i]);
            }
            spts.push(SparseTable::<bool, _>::new(&c, |x, y| x && y));
        }
        for _ in 0 .. k {
            let l: usize = get();
            let r: usize = get();
            let mut ok = false;
            if l < r {
                for i in 0 .. m {
                    ok |= spts[i].query(l - 1, r - 2);
                }
            } else {
                ok = true;
            }
            println!("{}", if ok { "Yes" } else { "No" });
        }
    }
}

fn main() {
    // In order to avoid potential stack overflow, spawn a new thread.
    let stack_size = 104_857_600; // 100 MB
    let thd = std::thread::Builder::new().stack_size(stack_size);
    thd.spawn(|| solve()).unwrap().join().unwrap();
}

File: codeforces/401/2c.cpp
#include <algorithm>
#include <bitset>
#include <cassert>
#include <cctype>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <deque>
#include <functional>
#include <iomanip>
#include <iostream>
#include <list>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <utility>
#include <vector>

#define REP(i,s,n) for(int i=(int)(s);i<(int)(n);i++)

using namespace std;
typedef long long int ll;
typedef vector<int> VI;
typedef vector<ll> VL;
typedef pair<int, int> PI;

/*
 * Sparse Table.
 * BiOp should be the type of a binary operator which is
 * associative, commutative and idempotent.
 * (For example, min and gcd satisfy them.)
 * Header Requirement: vector, cassert
 * Verified by: AtCoder ARC023 D (http://arc023.contest.atcoder.jp/submissions/960757)
 */
template<class T, class BiOp>
class SparseTable {
private:
  BiOp biop;
  std::vector<std::vector<T> > st;
  void create_sparse_table(int n, const std::vector<T> &lcp) {
    int h = 1;
    while ((1 << h) < n) {
      ++h;
    }
    st = std::vector<std::vector<T> >(h + 1, std::vector<T>(n));

    for (int i = 0; i < n; ++i) {
      st[0][i] = lcp[i];
    }
    for (int j = 1; j <= h; ++j) {
      for (int i = 0; i <= n - (1 << j); ++i) {
	st[j][i] = biop(st[j - 1][i], st[j - 1][i + (1 << (j-1))]);
      }
    }
  }
  /*
   * Reference: https://graphics.stanford.edu/~seander/bithacks.html#IntegerLogFloat
   * Please be aware that it only works well in case of 1 <= t <= 2^24.
   */
  static int top_bit(int t) {
    const float v = t; // find int(log2(v)), where v > 0.0 && finite(v) && isnormal(v)
    int c;         // 32-bit int c gets the result;
    
    c = *(const int *) &v;  // OR, for portability:  memcpy(&c, &v, sizeof c);
    return (c >> 23) - 127;
  }
public:
  /*
   * Initializes this sparse table. O(n log n) where n = ary.size().
   */
  SparseTable(BiOp biop, const std::vector<T> &ary): biop(biop) {
    create_sparse_table(ary.size(), ary);
  }
  /*
   * Computes biop(ary[f], ary[f+1], ..., ary[s]). O(1).
   * Note: the interval is inclusive.
   */
  T query(int f, int s) const {
    assert (f <= s);
    int diff = top_bit(s + 1 - f);
    return biop(st[diff][f], st[diff][s + 1 - (1 << diff)]);
  }
};


struct omax {
  int operator()(int x, int y) const {
    return max(x, y);
  }
};

int main(void){
  int n, m, k;
  cin >> n >> m;
  vector<VI> a(n, VI(m));
  REP(i, 0, n) {
    REP(j, 0, m) {
      cin >> a[i][j];
    }
  }
  cin >> k;
  VI l(k), r(k);
  REP(i, 0, k) {
    cin >> l[i] >> r[i];
    l[i]--; r[i]--;
  }
  if (m <= 320) {
    // Sparse Table Sol
    vector<SparseTable<int, omax> > spt;
    REP(i, 0, m) {
      VI ary(n - 1);
      REP(j, 0, n - 1) {
	ary[j] = a[j][i] <= a[j + 1][i] ? 0 : 1;
      }
      spt.push_back(SparseTable<int, omax>(omax(), ary));
    }
    REP(i, 0, k) {
      bool ans = false;
      REP(j, 0, m) {
	if (l[i] >= r[i] || spt[j].query(l[i], r[i] - 1) == 0) {
	  ans = true;
	}
      }
      cout << (ans ? "Yes" : "No") << endl;
    }
  } else {
    // O(n^2 m)
    vector<vector<bool> > ans(n, vector<bool>(n));
    REP(i, 0, n) {
      vector<bool> tmp(m, true);
      REP(j, i, n) {
	bool ok = false;
	REP(k, 0, m) {
	  ok = ok || tmp[k];
	}
	ans[i][j] = ok;
	if (j < n - 1) {
	  REP(k, 0, m) {
	    tmp[k] = tmp[k] && a[j][k] <= a[j + 1][k];
	  }
	}
      }
    }
    REP(i, 0, k) {
      cout << (ans[l[i]][r[i]] ? "Yes" : "No") << endl;
    }
  }
}
